File: _snapshots/zmq_i2c_backup_20260827_130857/gbt-sca-sw/sources/main/src/gbtsca_txhelper.cc
#include <stdint.h>
#include <stdlib.h>
#include <cstring>
#include <dirent.h>
#include <errno.h>
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>

#include <uhal/uhal.hpp>
#include <gbtsca_txhelper.h>

#include <vector>
#include <string>
#include <iostream>

using namespace uhal;

Tx_helper::Tx_helper(const uhal::HwInterface* ipbushw, const std::string& baseName) : node_(ipbushw->getNode(baseName+NODE_NAME)) {
    clear();
}

uint8_t Tx_helper::push(bool sendFifo, uint8_t transID)
{
    uint8_t transmittedTransID = 0;

    if(transID == 0 || transID == 255) // illegal values so we use these to indicate to auto set ID
    {
        data.transID = nextTransId_;
        transmittedTransID = nextTransId_;
        if(nextTransId_ >= 254) nextTransId_ = 1;
        else                    ++nextTransId_;
    }
    else
    {
        data.transID = transID;
        transmittedTransID = transID;
    }
  
    node_.getNode("address").write(data.address);
    node_.getNode("transID").write(data.transID);
    node_.getNode("channel").write(data.channel);
    node_.getNode("length").write(data.length);
    node_.getNode("command").write(data.command);
    node_.getNode("data0").write(data.data0);
    node_.getNode("data1").write(data.data1);
    node_.getNode("data2").write(data.data2);
    node_.getNode("data3").write(data.data3);

    node_.getClient().dispatch();
    if(sendFifo)
    {
        node_.getNode("fifo_fill").write(1);
        node_.getNode("fifo_go").write(1);
    }
    else
    {
        node_.getNode("fifo_fill").write(1);
    }
    node_.getClient().dispatch();

    return transmittedTransID;
}

void Tx_helper::send()
{
    node_.getNode("fifo_go").write(1);
    node_.getClient().dispatch();
    usleep(100);
    node_.getNode("fifo_fill").write(0);
    node_.getClient().dispatch();
}

void Tx_helper::softReset()
{
    node_.getNode("fifo_go").write(4);
    node_.getClient().dispatch();
    usleep(10);
    node_.getNode("fifo_go").write(0);
    node_.getClient().dispatch();
}


void Tx_helper::print()
{
    printf("Tx address: %8x\n", data.address);
    printf("Tx transId: %8x\n", data.transID);
    printf("Tx channel: %8x\n", data.channel);
    printf("Tx command:       %02x\n", data.command);
    printf("Tx length:  %8x\n", data.length);
    printf("Tx data:    %02x%02x%02x%02x\n", data.data3, data.data2, data.data1, data.data0);	
}

void Tx_helper::clear()
{
    nextTransId_ = 2;
    data.address = 0;
    data.transID = 1; 
    data.channel = 0; 
    data.length = 0; 
    data.command = 0; 
    data.data3 = 0; 
    data.data2 = 0; 
    data.data1 = 0; 
    data.data0 = 0; 
}

File: _snapshots/zmq_i2c_backup_20260827_130857/gbt-sca-sw/python/gbtsca_txhelper.py
"""Handles storing and sending messages for the gbtsca"""
import time


class tx_data:
    def __init__(self):
        self.address = 0
        self.transID = 1
        self.channel = 0
        self.length = 0
        self.command = 0
        self.data0 = 0
        self.data1 = 0
        self.data2 = 0
        self.data3 = 0


class Tx_helper:
    def __init__(self, ipbushw, baseName):
        self.NODE_NAME = "_GBT_SCA_tx"
        self.node = ipbushw.getNode(baseName+self.NODE_NAME)
        self.data = tx_data()
        self.clear()
        self.nextTransId = 2

    def push(self, sendFifo=False, transID=0):
        transmittedTransID = 0

        # 0 and 255 illegal transaction IDs, used to indicate auto set ID.
        if(transID == 0 or transID == 255):
            self.data.transID = self.nextTransId
            transmittedTransID = self.nextTransId
            if self.nextTransId >= 254:
                self.nextTransId = 1
            else:
                self.nextTransId = self.nextTransId+1
        else:
            self.data.transID = transID
            transmittedTransID = transID

        self.node.getNode("address").write(self.data.address)
        self.node.getNode("transID").write(self.data.transID)
        self.node.getNode("channel").write(self.data.channel)
        self.node.getNode("length").write(self.data.length)
        self.node.getNode("command").write(self.data.command)
        self.node.getNode("data0").write(self.data.data0)
        self.node.getNode("data1").write(self.data.data1)
        self.node.getNode("data2").write(self.data.data2)
        self.node.getNode("data3").write(self.data.data3)

        self.node.getClient().dispatch()
        if sendFifo:
            self.node.getNode("fifo_fill").write(1)
            self.node.getNode("fifo_go").write(1)
        else:
            self.node.getNode("fifo_fill").write(1)
        self.node.getClient().dispatch()

        return transmittedTransID

    def send(self):
        self.node.getNode("fifo_go").write(1)
        self.node.getClient().dispatch()
        time.sleep(0.1)
        self.node.getNode("fifo_fill").write(0)
        self.node.getClient().dispatch()

    def softReset(self):
        self.node.getNode("fifo_go").write(4)
        self.node.getClient().dispatch()
        time.sleep(0.01)
        self.node.getNode("fifo_go").write(0)
        self.node.getClient().dispatch()

    def txprint(self):
        print("Tx address: "+self.data.address)
        print("Tx transId: "+self.data.transID)
        print("Tx channel: "+self.data.channel)
        print("Tx command:       "+self.data.command)
        print("Tx length:  "+self.data.length)
        print("Tx data:    "+self.data.data3+" "+self.data.data2 +
              " "+self.data.data1+" "+self.data.data0)

    def clear(self):
        self.nextTransId_ = 2
        self.data.address = 0
        self.data.transID = 1
        self.data.channel = 0
        self.data.length = 0
        self.data.command = 0
        self.data.data3 = 0
        self.data.data2 = 0
        self.data.data1 = 0
        self.data.data0 = 0
